### PR TITLE
SOLR-15059: Improve query performance monitoring

### PR DESCRIPTION
**Pull request by thelabdude**
# Description

See JIRA: https://issues.apache.org/jira/browse/SOLR-15059 ... see a detailed description in the JIRA

# Solution

Improve the Grafana dashboard to include graphs for monitoring query performance.

# Tests

Manual testing of the Grafana dashboard in the browser while running query load.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

**Comment by MarcusSorealheis** [comment]
Is `stripWs` a term of art? Its definition is semi-obvious but will be even more obvious with a new name. The test files will live a long time and this PR is a key addition.

**Comment by epugh** [comment]
How does this compare/overlap with any of the dashboards on Grafana's site?  https://grafana.com/grafana/dashboards?search=solr

Specifically, the one published by @janhoy at https://grafana.com/grafana/dashboards/12456?

Be nice if we had one widely used and supported Grafana dashboard!

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    String metric = matched.group("METRIC");
+    if (metric == null) {
+      metric = unique;
+    }
+
+    // could be a simple field name or some kind of function here
+    if (!metric.contains("$")) {
+      if ("object.value".equals(metric)) {
+        metric = "$object.value"; // don't require the user to supply the $
```
Comment (madrob) [inline solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java:95]: Are we trying to be too helpful here? Or is this matching some existing spec?

**Comment by thelabdude** [comment]
@epugh If you diff the dashboard you linked to with the version in master (https://github.com/apache/lucene-solr/blob/master/solr/contrib/prometheus-exporter/conf/grafana-solr-dashboard.json) you'll see they are the same.

This PR just enhances that dashboard with query performance metrics and changes a few of the defaults in how null / zero series are handled.

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      throw new IllegalArgumentException("jq template must not be empty");
+    }
+    this.defaultType = defaultType != null ? defaultType : "GAUGE";
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public String getDefaultType() {
```
Comment (madrob) [inline solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java:61]: unused?

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    return defaultType;
+  }
+
+  public String getTemplate() {
```
Comment (madrob) [inline solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java:65]: unused

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsConfiguration.java [hunk 4]
```diff
@@ -105,17 +108,23 @@ public static MetricsConfiguration from(String resource) throws Exception {
   public static MetricsConfiguration from(Document config) throws Exception {
     Node settings = getNode(config, "/config/settings");
 
+    Map<String,MetricsQueryTemplate> jqTemplatesMap = null;
+    NodeList jqTemplates = (NodeList)(xpathFactory.newXPath()).evaluate("/config/jq-templates/template", config, XPathConstants.NODESET);
```
[the remark this answers is not in the document]
Comment (thelabdude) [reply solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsConfiguration.java]: The code is already using XPath, I'm not introducing it here?

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java (hunk 1 above)
Comment (thelabdude) [reply solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsQueryTemplate.java:95]: just trying to be helpful to keep the template syntax simpler ... e.g. I like {{count}} vs. {{$object.value.count}}

File: solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (thelabdude) [reply solr/contrib/prometheus-exporter/src/java/org/apache/solr/prometheus/exporter/MetricsConfiguration.java]: Also, this happens once during initialization of the Prom exporter, so efficiency of XPath isn't so much of a concern. A few extra millis (if that much) during init doesn't seem like a problem to me, esp. if it makes the code cleaner.

**Comment by thelabdude** [comment]
> Is `stripWs` a term of art? Its definition is semi-obvious but will be even more obvious with a new name. The test files will live a long time and this PR is a key addition.

@MarcusSorealheis Updated the name ;-)

**Comment by thelabdude** [comment]
Pinged you for a review @janhoy if you're curious. Also, would appreciate guidance on updating the default dashboard with these changes once they are back ported to 8.8. Thanks in advance ;-)

File: solr/contrib/prometheus-exporter/conf/solr-exporter-config.xml [hunk 5]
```diff
@@ -315,88 +477,22 @@
...
-            $object.value.count as $value |
-            {
-              name         : "solr_metrics_node_requests_total",
-              type         : "COUNTER",
-              help         : "See following URL: https://lucene.apache.org/solr/guide/metrics-reporting.html",
-              label_names  : ["category", "handler"],
-              label_values : [$category, $handler],
-              value        : $value
-            }
+            $jq:node(requests_total, select(.key | endswith(".local.requestTimes")), count)
```
Comment (dsmiley) [inline solr/contrib/prometheus-exporter/conf/solr-exporter-config.xml:486]: @thelabdude I noticed here you added a ".local." when it wasn't there before.  Why?  And for that matter, maybe we needn't bother publishing this particular metric at all; I'm skeptical of the utility.

Comment (thelabdude) [reply solr/contrib/prometheus-exporter/conf/solr-exporter-config.xml:486]: @dsmiley it's to support the query charts showing core-level query metrics vs. top-level distributed query metrics added in this PR. I like knowing if there's an imbalance of core-level query requests going to certain replicas or if the load across all of my replicas is balanced. So you're skeptical but you haven't said why exactly? You want to change, then change it.

Comment (thelabdude) [reply solr/contrib/prometheus-exporter/conf/solr-exporter-config.xml:486]: fwiw ~ have you actually looked at the charts added in this PR in Grafana with query load running? If there's a problem there, then let's fix it and move forward but rehashing old decisions seems unproductive at this point.

Comment (dsmiley) [reply solr/contrib/prometheus-exporter/conf/solr-exporter-config.xml:486]: I meant to comment on the "totalTime" metric w.r.t. it's usefulness; sorry for the confusion.  It's some massive number of course... it'd need to be divided by something else to be useful?  Also, totalTime is in nanoseconds lately!  https://issues.apache.org/jira/browse/SOLR-16073

I understand the overarching objective of top-level vs core-level -- makes sense. 
I'm a bit unclear on the distinction between the node level "$jq:node" metrics, and the "Local (non-distrib) query metrics", both of which are using ".local.".

RE Grafana; I haven't seen our official one in use live.  I use our own/custom one at work.

**Pull request merged**